open-wc/open-wc
# docs: refresh on controllerchange

When we deploy a new version of the docs site, sometimes users see the loading bar load for a long time. The reason for this is that a new service worker has been installed, the old assets on the server are replaced with new assets, and the old cache has been emptied by the new service worker.

**How to reproduce:**
- run `npm run vuepress:build`
- `cd _site && http-server`
- open the page in the browser
- close the page in the browser
- make a change to `/docs/guide/README.md`
- run `npm run vuepress:build` again
- `cd _site && http-server`
- open the page in browser: its now broken

The page is still on the old version of the app, and so it requests assets that no longer exist either in the cache or on the server.

The fix for this is to add a refresh whenever the `controllerchange` event is fired, indicating that a new service worker has taken over. The page will then refresh (which is near-instant, because the new service worker has already cached new assets), and there will no longer be broken/missing assets.

The check for `if (refreshing)` is there to avoid infinite loops if you have `update on reload` enabled in dev tools.

You've probably seen this pattern happen before, Twitter uses this as well.

If you want to read/understand more about why/how this problem is happening, see:
https://jovial-davinci-20c5fc.netlify.com/skip-waiting
